File: tank_combat.pl
#!/usr/bin/perl

package Tank;

use strict;
use warnings;
use Carp;

sub new {
    my $class = shift;
    my $self  = { @_ };
    croak "bad arguments" unless defined $self->{type} and defined $self->{weight} and defined $self->{armor_class}; 
    return bless $self, $class; 
}

sub tank_object {
    my $self = shift;
    return "$self->{type} $self->{weight} $self->{armor_class}";
}

package main;

my $tank1 = Tank->new( type => "Sherman",   weight => 10000, armor_class =>  5 );
my $tank2 = Tank->new( type => "Panzer",    weight => 15000, armor_class => 15 );
my $tank3 = Tank->new( type => "M1 Abrams", weight => 20000, armor_class => 12 );

sub get_lowest_armor_class {
    my @armor_classes; 
    for ( $tank1, $tank2, $tank3 ) { 
        push @armor_classes, $_->{ armor_class };
    }
    my @sorted = sort { $a <=> $b } @armor_classes;
    return $sorted[0];  
}

sub find_tank_with_lowest_armor_class {
    my @tanks;
    for ( $tank1, $tank2, $tank3 ) {
        if ( $_->{ armor_class } == get_lowest_armor_class() ) {
            return $_->{ type };
        }
    }
}

my $winning_tank_type = find_tank_with_lowest_armor_class(); 
sub get_winning_tank_weight {
    for ( $tank1, $tank2, $tank3 ) {
        if ( $_->{ type } eq $winning_tank_type ) {
            return $_->{ weight };
        }
    }
}

print $tank1->tank_object,  "\n";
print $tank2->tank_object,  "\n";
print $tank3->tank_object,  "\n";
my $winning_tank = find_tank_with_lowest_armor_class(); 
my $winning_tank_weight = get_winning_tank_weight(); 
print "winning tank is: $winning_tank\n"; 
print "winning tank weight is: $winning_tank_weight\n"; 

File: TankCombat.java
public class TankCombat {
    public static void main( String[] args ) {
        Weapon tank1 = new Weapon( "Sherman",    10000, 140, 20 );
        Weapon tank2 = new Weapon( "Panzer" ,    15000, 120, 13 );
        Weapon tank3 = new Weapon( "M1 Abrams" , 20000, 180,  3 );

        System.out.println( tank1 ); 
        System.out.println( tank2 );
        System.out.println( tank3 ); 
        System.out.println( tank3.armor_class ); 

        int armor_class_array[] = new int[3];
        armor_class_array[0] = tank1.armor_class;
        armor_class_array[1] = tank2.armor_class;
        armor_class_array[2] = tank3.armor_class;
 
        int lowest_armor_class = armor_class_array[0];
        for ( int i = 0; i < 3; i++ ) {
            if ( armor_class_array[i] < lowest_armor_class ) {
                lowest_armor_class = armor_class_array[i];
            }
        }

        for ( int i = 0; i < 3; i++ ) {
            if ( armor_class_array[i] == lowest_armor_class ) {
                if ( i == 0 ) {
                    System.out.printf( "The winning tank is: %s\n", tank1.weapon_name ); 
                } else if ( i == 1 ) {
                    System.out.printf( "The winning tank is: %s\n", tank2.weapon_name ); 
                } else if ( i == 2 ) {
                    System.out.printf( "The winning tank is: %s\n", tank3.weapon_name ); 
                }
            }
        }
        System.out.printf( "the winning tank has armor class: %d\n", lowest_armor_class ); 
    }
}

